Backend/FcEvaluate.hs

```diff
@@ -0,0 +1,143 @@
+  FcPgmDataDecl _ p                  -> collapseProgram p
+  FcPgmValDecl (FcValBind x ty e1) p -> FcTmLet x ty e1 $ collapseProgram p
+
+-- | Evaluate an expression (small-step).
+smallStep :: FcTerm 'Fc -> EvM (Maybe (FcTerm 'Fc))
+-- Values (see why we need or-patterns :P)
+smallStep (FcTmAbs {})     = pure Nothing
+smallStep (FcTmTyAbs {})   = pure Nothing
+smallStep (FcTmUnit {})    = pure Nothing
+smallStep (FcTmDataCon {}) = pure Nothing
+
+-- Normal forms (but not values)
+smallStep (FcTmVar {})   = pure Nothing
+smallStep (FcTmERROR {}) = pure Nothing
+
+-- Error propagation
+smallStep (FcTmApp    e1@(FcTmERROR {}) _) = pure $ Just e1
+smallStep (FcTmTyApp  e1@(FcTmERROR {}) _) = pure $ Just e1
+smallStep (FcTmCaseFc e1@(FcTmERROR {}) _) = pure $ Just e1
+
+-- Useful cases (substitution)
+smallStep (FcTmApp (FcTmAbs x _ body) e2) = Just <$> do
+  newe <- freshenLclBndrs e2
+  pure $ substVar x newe body
+smallStep (FcTmTyApp (FcTmTyAbs a body) ty) = Just <$> do
+  newty <- freshenLclBndrs ty
+  pure $ substVar a newty body
+smallStep (FcTmLet x ty e1 e2) = do
+  newe1 <- freshenLclBndrs $ FcTmLet x ty e1 e1
+  pure <$> Just $ substVar x newe1 e2
+smallStep (FcTmCaseFc scr alts)
+  | FcTmUnit <- scr
+  = matchTheUnit alts
+  | Just (dc, args) <- userDefinedDataConAppMaybe scr
+  = matchTheAlts dc args alts
+
+-- Congruences
+smallStep (FcTmApp e1 e2) = smallStep e1 >>= \case
+  Just e1' -> pure $ Just $ FcTmApp e1' e2
+  Nothing  -> pure Nothing
+smallStep (FcTmTyApp e1 ty) = smallStep e1 >>= \case
+  Just e1' -> pure $ Just $ FcTmTyApp e1' ty
+  Nothing  -> pure Nothing
+smallStep (FcTmCaseFc scr alts) = smallStep scr >>= \case
+  Just scr' -> pure $ Just $ FcTmCaseFc scr' alts
+  Nothing   -> pure $ Nothing
+
+matchTheUnit :: [FcAlt 'Fc] -> EvM (Maybe (FcTerm 'Fc))
+matchTheUnit []                              = error "<unit impossible: pm compiler failed!>"
+matchTheUnit ((FcAltFc FcUnitPat     rhs):_) = pure $ Just rhs
+matchTheUnit ((FcAltFc (FcConPat {}) _  ):_) = error "<unit impossible: elab failed!>"
+
+matchTheAlts :: FcDataCon -> [FcTerm 'Fc] -> [FcAlt 'Fc] -> EvM (Maybe (FcTerm 'Fc))
+matchTheAlts _dc _args []                           = error "<impossible: pm compiler failed!>"
+matchTheAlts _dc _args ((FcAltFc FcUnitPat _rhs):_) = error "<impossible: elab failed!>"
+matchTheAlts  dc  args ((FcAltFc (FcConPat dc' xs) rhs):rest)
+  | dc == dc' = if
+      | length args /= length xs -> error "<impossible: elab failed!>"
+      | otherwise -> do
+          sub <- buildFcTmSubst . zipExact xs <$> mapM freshenLclBndrs args
+          pure $ Just $ substFcTmInTm sub rhs
+  | otherwise = matchTheAlts dc args rest
+
+
+-- collapseProgram :: FcProgram 'Fc -> FcTerm 'Fc
+-- smallStep :: MonadUnique m => FcTerm 'Fc -> m (Maybe (FcTerm 'Fc))
+
+groundTerm :: FcTerm 'Fc -> FcTerm 'Fc
+groundTerm (FcTmTyAbs ty t1) = FcTmTyApp (FcTmTyAbs ty (groundTerm t1)) FcTyUnit
+groundTerm t                 = t
```

Comment:
I think this one also deserves a comment about how it works. Also, the comment should mention that this is only patched grounding. For example
```haskell
let f = <expr1> in /\a. <expr2>
```
will not be grounded, but this one
```haskell
/\a. let f = <expr1> in <expr2>
```
will (even though both of them could initially have the exact same type). The reason that this problem does not arise in our case is because WE KNOW that our type inference and elaboration algorithm produces type abstraction only at the type level. This is the reason that `groundTerm` suffices in our case, but this should still be explained in a comment.